index.ts

```
@@ -29,11 +31,20 @@ export async function run(args: { [k: string]: boolean | string }): Promise<void
     // todo: Update usage() to provide a hint that wallet is missing/malformed.
     usage(""); // no return
   } else {
-    await cmds[cmd](logger);
+    do {
+      try {
+        await cmds[cmd](logger);
+      } catch (error) {
+        // eslint-disable-next-line no-process-exit
+        if (await processCrash(logger, cmd, config.pollingDelay, error)) process.exit(1);
```

you can just re-throw the error rather than forcibly exiting the process
EDIT: i see now that this pattern was already set by previous code

---

Do you think we should just throw the error instead of process.exit-ing? Seems like we should avoid the low level process.X calls if possible

---

I think the process.exit call is important at the top level to avoid "unhandled promise rejection" printouts, but I think the calling function (see line 56) does that for us. I would think rethrowing in this case would be sufficient here, right?

---

> very subtle changes, its hard to know as the reviewer if things are working as intended. maybe you could provide how you verified the changes work? otherwise the code seems fine, though my gut says the existing architecture looks fragile.

Just to give feedback on the question about how it's been tested - I've forced exceptions by inserting a 'throw "XXX";' statement into the for(;;) loops of each "command" (i.e. runRelayer()).

---

> Do you think we should just throw the error instead of process.exit-ing? Seems like we should avoid the low level process.X calls if possible

> I think the process.exit call is important at the top level to avoid "unhandled promise rejection" printouts, but I think the calling function (see line 56) does that for us. I would think rethrowing in this case would be sufficient here, right?

Per the latest change, I've removed the 2nd-tier "processCrash()" calls. So it's now only handled in the top-level index.ts. I think this is cleaner and makes the behaviour more obvious.